### ISSUE TITLE
bug(#480): SD1.5 VAE forward (encode + decode) diverges from diffusers — img2img collapses to yellow

### ISSUE THREAD

**Issue by dahai80** [dahai80/fusion-mlx#482]
## Summary

The SD1.5 VAE (`fusion_mlx/image/sdxl/vae.py`, shared `SDXLVAE` reused via `SD15VAE`) produces **wrong outputs on both `encode` and `decode`** compared to diffusers `AutoencoderKL` with the **same weights and the same input/latent**. This corrupts SD1.5 img2img (2nd pass of `2_pass_txt2img` hires-fix) and means SD1.5 txt2img outputs are also not faithful to the reference pipeline.

Discovered while running ComfyUI's canonical `2_pass_txt2img` example against the fusion-mlx backend (issue #480). Pass-1 txt2img (512×512) yields a plausible-looking image, but pass-2 img2img (encode real image → noise → denoise → decode) collapses to a saturated-yellow image (R≈0.96, G≈0.99, **B≈0.00**).

## Reproduction

```python
import os, numpy as np, torch
from PIL import Image
from diffusers import AutoencoderKL
import mlx.core as mx
from fusion_mlx.image.sd15.generate import SD15Pipeline, _pil_to_nchw

snap = sorted(os.listdir(os.path.expanduser(
    "~/.fusion-mlx/models/models--runwayml--stable-diffusion-v1-5/snapshots")))[0]
ref = AutoencoderKL.from_pretrained(
    os.path.expanduser(f"~/.fusion-mlx/models/models--runwayml--stable-diffusion-v1-5/snapshots/{snap}/vae")
).to("cpu").eval()

img = Image.new("RGB", (512, 512))  # any real image
x_neg = torch.from_numpy(np.array(img).astype(np.float32)/255*2-1).permute(2,0,1)[None]

# --- ENCODE ---
with torch.no_grad():
    lat_ref = (ref.encode(x_neg).latent_dist.sample()*0.18215).numpy()[0]   # [4,64,64]
p = SD15Pipeline(); p._load_all()
lat_mlx = np.array(p.vae.encode(_pil_to_nchw(img, 512, 512)).astype(mx.float32))[0]

# --- DECODE the SAME diffusers latent ---
with torch.no_grad():
    dec_ref = (ref.decode(torch.from_numpy(lat_ref)/0.18215).sample.numpy()[0]/2+0.5).clip(0,1)
dec_mlx = (np.array(p.vae.decode(mx.array(lat_ref)).astype(mx.float32))[0]/2+0.5).clip(0,1)
```

## Observed (a real 512×512 photo as input)

**Encode** — per-channel latent mean (`lat_ref` vs `lat_mlx`):
| ch | diffusers | mlx | Δ |
|----|-----------|-----|---|
| 0  | +0.960    | -1.130 | -2.09 |
| 1  | -0.531    | -1.598 | -1.07 |
| 2  | -0.007    | -0.081 | -0.07 |
| 3  | +0.600    | -0.816 | -1.42 |

Latent std matches (~0.70 vs ~0.68) but the **per-channel mean is badly off and non-uniform** — no channel permutation aligns them.

**Decode** (feeding the *same* diffusers-encoded latent to both decoders):
- diffusers → `[0.531, 0.411, 0.322]` (faithful to input)
- mlx → `[0.268, 0.116, 0.105]` (dark / wrong), correlation **0.57**

So both directions diverge; it is not an encode-only or decode-only issue.

## What is NOT the cause (verified component-by-component vs PyTorch)

Each shared primitive was unit-tested against torch and matches (corr 1.0):
- `Conv2d` (OIHW→OHWI transpose, stride/pad) ✅
- `GroupNorm` (32 groups, NHWC) ✅
- `Upsample` (broadcast-reshape nearest 2x + conv) ✅
- `mlx.core.fast.scaled_dot_product_attention` (single-head `[B,1,N,D]`, scale=1/√D) ✅

Weight loading reports **strict, 0 missing** for the SD1.5 VAE (248/248 keys; legacy `query/key/value/proj_attn`→`to_q/to_k/to_v/to_out.0` remap applied). So it is not a missing-weight problem.

## Suspects

Since every primitive is correct yet the composed encoder/decoder both diverge, the bug is likely in a **structural wiring** detail shared by both paths:
- `ResnetBlock` (`conv_shortcut` 1×1 conv path / residual add) — not yet unit-tested in isolation
- `MidBlock` attention wiring or the decoder `reversed_mult` / `up_blocks` channel progression
- A subtle NHWC↔NCHW transpose or `conv_shortcut` padding (`Conv2d(k=1, padding=0)` default has `padding=1`!) — **note `Conv2d.__init__` defaults `padding=1`, but `conv_shortcut=Conv2d(in_ch,out_ch,k=1)` relies on the k=1 branch; `Upsample`/`Downsample` also construct `Conv2d` with defaults — worth checking the 1×1 shortcut actually uses padding=0.**

## Impact

- SD1.5 img2img (hires-fix 2nd pass) is unusable: output collapses to saturated yellow.
- SD1.5 txt2img produces plausible but **non-faithful** images vs the reference SD1.5 pipeline (decoder maps UNet latents differently than diffusers would).
- Blocks the `2_pass_txt2img` ComfyUI example as an Apple-Silicon-vs-5090 comparison point (#480).

## Environment
- fusion-mlx `feat/sd15-pipeline-img2img-480` branch
- MLX (mlx.core), Apple Silicon
- diffusers 0.39.0 (reference)
- Model: `runwayml/stable-diffusion-v1-5` VAE (`vae/diffusion_pytorch_model.safetensors`)

**Comment by dahai80** [comment]
Update: unit-tested `ResnetBlock` (norm1→silu→conv1→norm2→silu→conv2 + 1×1 conv_shortcut residual) in isolation vs PyTorch — **corr 1.0, max abs diff 2e-4**. So ResnetBlock is also correct.

Component verification status (all vs PyTorch, corr 1.0): Conv2d ✅, GroupNorm ✅, Upsample (nearest) ✅, SDPA ✅, ResnetBlock ✅.

This narrows the bug to the **decoder/encoder structural wiring** (MidBlock attention placement, `VAEDecoder` `reversed_mult`/up_blocks channel progression, conv_in channel counts, or the conv_norm_out/conv_out assembly) — i.e. how the verified primitives are *composed*, not the primitives themselves.
